File: python/amqp_producer.py
"""
AMQP Producer (Python): reads FIXM-inspired JSON, applies S/MIME,
and publishes to an AMQP 1.0 queue.
Demonstrates SWIM-TIYP-0112 over AMQP with Message Security binding.
"""

import os
import sys
import argparse

from smime_helper import (
    load_private_key, load_certificate,
    sign, sign_and_encrypt
)


def main():
    parser = argparse.ArgumentParser(description="SWIM S/MIME AMQP Producer (Python)")
    parser.add_argument("--certs-dir", default=os.path.join(os.path.dirname(__file__), "..", "certs"))
    parser.add_argument("--payload", default=os.path.join(os.path.dirname(__file__), "..", "payload", "sample-flight.json"))
    parser.add_argument("--broker", default="localhost:5672")
    parser.add_argument("--queue", default="swim.flight.data")
    parser.add_argument("--mode", choices=["sign", "sign-encrypt"], default="sign")
    args = parser.parse_args()

    print("=== SWIM S/MIME AMQP Producer (Python) ===")
    print(f"Mode: {args.mode}")
    print(f"Broker: {args.broker}")
    print(f"Queue: {args.queue}")

    with open(args.payload, "rb") as f:
        payload = f.read()
    print(f"Payload loaded: {len(payload)} bytes")

    producer_key = load_private_key(os.path.join(args.certs_dir, "producer.pfx"), "changeit")
    producer_cert = load_certificate(os.path.join(args.certs_dir, "producer.pfx"), "changeit")

    if args.mode == "sign-encrypt":
        consumer_cert = load_certificate(os.path.join(args.certs_dir, "consumer.crt"))
        smime_data, content_type = sign_and_encrypt(
            payload, "application/json",
            producer_key, producer_cert, consumer_cert)
        print("Message signed and encrypted")
    else:
        smime_data, content_type = sign(
            payload, "application/json",
            producer_key, producer_cert)
        print("Message signed")

    print(f"S/MIME Content-Type: {content_type}")

    try:
        from proton import Message
        from proton.handlers import MessagingHandler
        from proton.reactor import Container

        class SmimeSender(MessagingHandler):
            def __init__(self, broker, queue, smime_data, content_type):
                super().__init__()
                self.broker = broker
                self.queue = queue
                self.smime_data = smime_data
                self.content_type = content_type
                self.sent = False

            def on_start(self, event):
                conn = event.container.connect(self.broker,
                                                user="admin", password="admin",
                                                allowed_mechs="PLAIN")
                self.sender = event.container.create_sender(conn, self.queue)

            def on_sendable(self, event):
                if not self.sent:
                    msg = Message()
                    msg.body = self.smime_data
                    msg.content_type = self.content_type
                    # SWIM-TIYP-0112: MIME-Version as application property (with hyphen per spec)
                    msg.properties = {"MIME-Version": "1.0"}
                    event.sender.send(msg)
                    self.sent = True
                    print(f"Message sent to AMQP queue: {self.queue}")
                    event.connection.close()

        Container(SmimeSender(args.broker, args.queue, smime_data, content_type)).run()

    except ImportError:
        print("ERROR: python-qpid-proton is not available.")
        print("Install it via: pip install python-qpid-proton")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: java/src/main/java/com/swim/smime/AmqpProducer.java
package com.swim.smime;

import org.apache.qpid.protonj2.client.*;

import java.nio.file.Files;
import java.nio.file.Path;
import java.security.PrivateKey;
import java.security.cert.X509Certificate;

/**
 * AMQP Producer: reads FIXM-inspired JSON, applies S/MIME sign+encrypt,
 * and publishes to an AMQP 1.0 queue using native AMQP 1.0 client (ProtonJ2).
 * Demonstrates SWIM-TIYP-0112 over AMQP with Message Security binding.
 */
public class AmqpProducer {

    public static void main(String[] args) throws Exception {
        String certsDir = args.length > 0 ? args[0] : "../certs";
        String payloadFile = args.length > 1 ? args[1] : "../payload/sample-flight.json";
        String brokerHost = args.length > 2 ? args[2] : "localhost";
        int brokerPort = args.length > 3 ? Integer.parseInt(args[3]) : 5672;
        String queueName = args.length > 4 ? args[4] : "swim.flight.data";
        String mode = args.length > 5 ? args[5] : "sign-encrypt";

        System.out.println("=== SWIM S/MIME AMQP Producer (Java) ===");
        System.out.println("Mode: " + mode);
        System.out.println("Broker: " + brokerHost + ":" + brokerPort);
        System.out.println("Queue: " + queueName);

        byte[] payload = Files.readAllBytes(Path.of(payloadFile));
        System.out.println("Payload loaded: " + payload.length + " bytes");

        PrivateKey producerKey = SmimeHelper.loadPrivateKey(
                Path.of(certsDir, "producer.pfx").toString(), "changeit");
        X509Certificate producerCert = SmimeHelper.loadCertificateFromPfx(
                Path.of(certsDir, "producer.pfx").toString(), "changeit");

        SmimeHelper.SmimeResult result;

        if ("sign-encrypt".equals(mode)) {
            X509Certificate consumerCert = SmimeHelper.loadCertificate(
                    Path.of(certsDir, "consumer.crt").toString());
            result = SmimeHelper.signAndEncrypt(payload, "application/json",
                    producerKey, producerCert, consumerCert);
            System.out.println("Message signed and encrypted");
        } else {
            result = SmimeHelper.sign(payload, "application/json",
                    producerKey, producerCert);
            System.out.println("Message signed");
        }

        String contentType = result.contentType().replaceAll("\\r?\\n[ \\t]+", " ");
        System.out.println("S/MIME Content-Type: " + contentType);

        // Send via native AMQP 1.0 (ProtonJ2)
        Client client = Client.create();
        ConnectionOptions connOpts = new ConnectionOptions();
        connOpts.user("admin");
        connOpts.password("admin");

        try (Connection connection = client.connect(brokerHost, brokerPort, connOpts);
             Sender sender = connection.openSender(queueName)) {

            Message<byte[]> message = Message.create(result.data());

            // SWIM-TIYP-0112: content-type as AMQP core property
            message.contentType(contentType);

            // SWIM-TIYP-0112: MIME-Version as application property (with hyphen)
            message.property("MIME-Version", "1.0");

            sender.send(message);
            System.out.println("Message sent to AMQP queue: " + queueName);
        }
    }
}
